sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,233 @@
+ */
+typedef enum dif_spi_device_result {
+  /**
+   * Indicates a successful operation.
+   */
+  kDifSpiDeviceResultOk = 0,
+  /**
+   * Indicates a failed precondition on the function arguments.
+   */
+  kDifSpiDeviceResultBadArg,
+} dif_spi_device_result_t;
+
+/**
+ * The length of the SPI device SRAM, in bytes.
+ *
+ * Useful for initializing FIFO lengths.
+ */
+extern const uint16_t kDifSpiDeviceSramLen;
+
+/**
+ * Configuration for initializing a SPI device.
+ */
+typedef struct dif_spi_device_config {
+  mmio_region_t base_addr;
+  dif_spi_device_edge_t clock_polarity;
+  dif_spi_device_edge_t data_phase;
+  dif_spi_device_bit_order_t tx_order;
+  dif_spi_device_bit_order_t rx_order;
+  uint16_t rx_fifo_timeout;  // uint12_t
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_len;
+} dif_spi_device_config_t;
+
+/**
+ * State for a particular SPI device.
+ *
+ * Its member variables should be considered private, and are only provided so
+ * that callers can allocate it.
+ */
+typedef struct dif_spi_device {
+  mmio_region_t base_addr;
+  uint16_t rx_fifo_base;
+  uint16_t rx_fifo_len;
+  uint16_t tx_fifo_base;
+  uint16_t tx_fifo_len;
+} dif_spi_device_t;
+
+/**
+ * Initializes a SPI device with the given configuration.
+ *
+ * @param config the configuration to initialize with.
+ * @param spi out param for the initialized device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_init(dif_spi_device_config_t config,
+                                            dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi);
+
+/**
+ * Issues an "abort" to the given SPI device, causing all in-progress IO to
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_abort(const dif_spi_device_t *spi);
+
+/**
+ * Checks whether a particular interrupt bit has been set.
+ *
+ * @param spi a SPI device.
+ * @param type which IRQ type to check.
+ * @param flag out param for the interrupt bit.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_get(const dif_spi_device_t *spi,
+                                               dif_spi_device_irq_type_t type,
+                                               bool *flag_out);
+
+/**
+ * Clears all active interrupt bits.
+ *
+ * @param spi a SPI device.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_clear_all(
+    const dif_spi_device_t *spi);
+
+/**
+ * Toggles a particular interrupt enable bit to enabled or disabled.
+ *
+ * @param spi a SPI device.
+ * @param type which IRQ type to toggle.
+ * @param state the state to update the bit to.
+ * @return the result of the operation.
+ */
+dif_spi_device_result_t dif_spi_device_irq_toggle(
```

Comment:
I think `dif_spi_device_irq_enable` is a better name, it is obvious from parameters and description that it also can disable an IRQ.

Reply:
Sure.